### PR TITLE
Large Chunk (Mmap) Overlapping

### PR DESCRIPTION
**Pull request by mdulin2**
Hey there, 

I had been playing around with very large allocations in GLibC and realized they are handled in an entirely different way than normal chunks in Malloc. Because of this, exploiting memory corruptions on mmapped chunks is a totally different ballgame.

This is a basic proof of concept that is similar to the overlapping chunks attack but just on mmap chunks. Because no POC's within this repository existed for anything on mmapped chunks (which operate completely differently), I thought it would be worthwhile to add some mmap chunk corruption into this repo. For the main difference between the attacks, review the source code that is coming along this request; it has a fairly detailed explanation on why mmap chunks are different and how this changes exploitation. 

There are several other techniques with mmap chunks that can be interesting, with a large chunk of them written up in http://tukan.farm/2016/07/27/munmap-madness/.

### CODE REVIEW

**Comment by mdulin2** [comment]
I have also verified this to work on versions 2.23, 2.26 and 2.29 of GLibC.

**Comment by Kyle-Kyle** [comment]
Very interesting idea. Basically, it turns a size overwritten(could be caused by many reasons) to chunk overlapping but at mmap level.
Our community currently pays very little attention to the very large allocations situation. I think this could be a good start to bring out attention to the very large bins.
But before that, can you resolve the review posted above? Thanks!

**Comment by mdulin2** [comment]
Yep, exactly what we are doing here. 

Thanks! I am excited to get some research for this type of exploitation more out into the open. There is a lot of research still to be done in this area still. 

@Kyle-Kyle I do not see a review on this pull request. Am I missing something?

**Comment by Kyle-Kyle** [comment]
Apparently I made a mistake and didn't submit the review lol. Sorry about that.

**Comment by mdulin2** [comment]
Ahh, I see haha. Yeah, I'll go through the review and resolve the comments. Thanks!

File: glibc_2.26/mmap_overlapping_chunks.c [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	of allocating it on the normal heap. This is determined by the mmap_threshold var.
+	Instead of the normal logic for getting a chunk, the system call *Mmap* is 
+	used. This allocates a section of virtual memory and gives it back to the user. 
+
+	Similarly, the freeing process is going to be different. Instead 
+	of a free chunk being given back to a bin or to the rest of the heap,
+	another syscall is used: *Munmap*. This takes in a pointer of a previously 
+	allocated Mmap chunk and releases it back to the kernel. 
+
+	Mmap chunks have special bit set on the size metadata: the second bit. If this
```
[the remark this answers is not in the document]
Comment (mdulin2) [reply glibc_2.26/mmap_overlapping_chunks.c:25]: According to this article, it is the second bit: https://sploitfun.wordpress.com/2015/02/10/understanding-glibc-malloc/ .Additionally, throughout the code I use `size | 0x2` in order to add this bit. The third bit is the `is not main arena` bit.

**Comment by Kyle-Kyle** [comment]
It looks good now. Thank you so much!

**Pull request merged**